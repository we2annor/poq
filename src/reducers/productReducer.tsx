import _ from "lodash";
import {
  FetchProductAction,
  DeleteProductAction,
  FETCH_PRODUCTS,
  DELETE_PRODUCT,
} from "../Actions/types";

export interface ProductState {
  products: Product[];
}

export interface Product {
  productId: number;
  name: string;
  description: string;
  price: number;
  priceWas: number;
  available: boolean;
  quantity: number;
  lowOnStock: boolean;
  promotionBadge: string;
  imageUrl: string;
}
const initialState: ProductState = {
  products: [],
};

export const productReducer = (
  state: ProductState = initialState,
  action: FetchProductAction | DeleteProductAction
) => {
  switch (action.type) {
    case FETCH_PRODUCTS:
      return { ...state.products, ..._.mapKeys(action.payload, "productId") };
    case DELETE_PRODUCT:
      const newProducts = Object.values(state);
      const filteredList = newProducts.filter(
        (product) => product.productId !== action.payload
      );
      return {
        ...filteredList,
      };
    default:
      return state;
  }
};
