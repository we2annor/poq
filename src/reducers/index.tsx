import { combineReducers } from "redux";
import { productReducer } from "./productReducer";

export type RootState = ReturnType<typeof rootReducer>;
export const rootReducer = combineReducers({ products: productReducer });
