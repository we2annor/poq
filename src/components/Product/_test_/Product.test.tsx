import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import { rootReducer } from "../../../reducers";
import ProductItem from "../Product";
import { Product } from "../../../reducers/productReducer";

afterEach(cleanup);

const product: Product = {
  productId: 1,
  name: "Dress 1",
  description: "A nice dredd",
  price: 12,
  priceWas: 15,
  available: true,
  quantity: 127,
  lowOnStock: false,
  promotionBadge: "30% OFF",
  imageUrl: "url/test.com",
};

const getSelectedProducts = (selectedProduct: HTMLInputElement) => {
  console.log(selectedProduct);
};

describe("When ProdcutItem Mounts", () => {
  test("Should Renders successfully", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    render(
      <Provider store={store}>
        <ProductItem
          product={product}
          getSelectedProducts={getSelectedProducts}
        />
      </Provider>
    );
  });

  test("Matches snapshot", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    const tree = renderer.create(
      <Provider store={store}>
        <ProductItem
          product={product}
          getSelectedProducts={getSelectedProducts}
        />
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
