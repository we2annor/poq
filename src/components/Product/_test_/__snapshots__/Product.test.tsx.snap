// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Address Mounts Matches snapshot 1`] = `
<div
  className="product"
>
  <span
    className="product__checkBox"
  >
    <input
      checked={false}
      className="remove-checkbox"
      id="1"
      name="1"
      onChange={[Function]}
      type="checkbox"
    />
  </span>
  <div
    className="product__img-wrapper"
  >
    <img
      alt="Dress 1"
      className="sc-bdfBwQ iBymIc"
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
    />
    <div
      className="product__promotionBadge"
    >
      30% OFF
    </div>
  </div>
  <div
    className="product__info"
  >
    <div
      className="product__info--name"
    >
      Dress 1
    </div>
    <div
      className="product__info--price"
    >
      £
      12
       
      <span>
        £
        15
      </span>
    </div>
    <div>
      <span
        className="in-stock"
      >
        127
         in stock
      </span>
    </div>
    <span>
      <span
        className="out-of-stock"
      >
        Out of stock
      </span>
    </span>
  </div>
</div>
`;

exports[`When ProdcutItem Mounts Matches snapshot 1`] = `
<div
  className="product"
>
  <span
    className="product__checkBox"
  >
    <input
      checked={false}
      className="remove-checkbox"
      id="1"
      name="1"
      onChange={[Function]}
      type="checkbox"
    />
  </span>
  <div
    className="product__img-wrapper"
  >
    <img
      alt="Dress 1"
      className="sc-bdfBwQ iBymIc"
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
    />
    <div
      className="product__promotionBadge"
    >
      30% OFF
    </div>
  </div>
  <div
    className="product__info"
  >
    <div
      className="product__info--name"
    >
      Dress 1
    </div>
    <div
      className="product__info--price"
    >
      £
      12
       
      <span>
        £
        15
      </span>
    </div>
    <div>
      <span
        className="in-stock"
      >
        127
         in stock
      </span>
    </div>
    <span>
      <span
        className="out-of-stock"
      >
        Out of stock
      </span>
    </span>
  </div>
</div>
`;
